### PR TITLE
emscripten: 3.1.10 -> 3.1.14

### PR DESCRIPTION
**Pull request by willcohen**
###### Description of changes

Rebase of #175358 to current master, re-enabling builds for darwin as well. @yorickvP 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by willcohen** [comment]
Currently fails on linux:

```
_info.json" for subsequent builds)
cache:INFO:  - ok
error: undefined symbol: main/__main_argc_argv (referenced by top-level compiled C/C++ code)
warning: Link with `-sLLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-sERROR_ON_UNDEFINED_SYMBOLS=0`
warning: _main may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
warning: To build in STANDALONE_WASM mode without a main(), use emcc --no-entry
Error: Aborting compilation due to previous errors
emcc: error: '/nix/store/lmh4dfiai7bs3ryyvvj4zfzbs7i3rzq6-nodejs-16.15.0/bin/node /nix/store/541iiim5rp8irjfdv2gl0ax5d37k816q-emscripten-3.1.14/share/emscripten/src/compiler.js /build/tmply98mnch.json' failed (returned 1)
error: builder for '/nix/store/35l2nn00962z6qpyjp8q0ncm2khvbvga-emscripten-3.1.14.drv' failed with exit code 1;
       last 10 log lines:
       > cache:INFO:  - ok
       > cache:INFO: generating system asset: sysroot/lib/wasm32-emscripten/struct_info.json... (this will be cached in "/nix/store/541iiim5rp8irjfdv2gl0ax5d37k816q-emscripten-3.1.14/share/emscripten/cache/sysroot/lib/wasm32-emscripten/struct_info.json" for subsequent builds)
       > cache:INFO:  - ok
       > error: undefined symbol: main/__main_argc_argv (referenced by top-level compiled C/C++ code)
       > warning: Link with `-sLLD_REPORT_UNDEFINED` to get more information on undefined symbols
       > warning: To disable errors for undefined symbols use `-sERROR_ON_UNDEFINED_SYMBOLS=0`
       > warning: _main may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
       > warning: To build in STANDALONE_WASM mode without a main(), use emcc --no-entry
       > Error: Aborting compilation due to previous errors
       > emcc: error: '/nix/store/lmh4dfiai7bs3ryyvvj4zfzbs7i3rzq6-nodejs-16.15.0/bin/node /nix/store/541iiim5rp8irjfdv2gl0ax5d37k816q-emscripten-3.1.14/share/emscripten/src/compiler.js /build/tmply98mnch.json' failed (returned 1)
       For full logs, run 'nix log /nix/store/35l2nn00962z6qpyjp8q0ncm2khvbvga-emscripten-3.1.14.drv'.
```

**Comment by willcohen** [comment]
Result of `nixpkgs-review pr 178620` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>faustStk</li>
    <li>magnetophonDSP.ConstantDetuneChorus</li>
  </ul>
</details>
<details>
  <summary>34 packages built:</summary>
  <ul>
    <li>binaryen</li>
    <li>emscripten</li>
    <li>emscriptenPackages.json_c</li>
    <li>emscriptenPackages.libxml2</li>
    <li>emscriptenPackages.xmlmirror</li>
    <li>emscriptenPackages.zlib</li>
    <li>faust</li>
    <li>faust2alqt</li>
    <li>faust2alsa</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jack</li>
    <li>faust2jackrust</li>
    <li>faust2jaqt</li>
    <li>faust2ladspa</li>
    <li>faust2lv2</li>
    <li>faustPhysicalModeling</li>
    <li>faustlive</li>
    <li>kapitonov-plugins-pack</li>
    <li>magnetophonDSP.CharacterCompressor</li>
    <li>magnetophonDSP.CompBus</li>
    <li>magnetophonDSP.LazyLimiter</li>
    <li>magnetophonDSP.MBdistortion</li>
    <li>magnetophonDSP.RhythmDelay</li>
    <li>magnetophonDSP.VoiceOfFaust</li>
    <li>magnetophonDSP.faustCompressors</li>
    <li>magnetophonDSP.pluginUtils</li>
    <li>magnetophonDSP.shelfMultiBand</li>
    <li>mooSpace</li>
    <li>open-music-kontrollers.mephisto</li>
    <li>python310Packages.filecheck</li>
    <li>python39Packages.filecheck</li>
    <li>tambura</li>
    <li>tinygo</li>
  </ul>
</details>

**Comment by willcohen** [comment]
The bump to 3.1.14 works when `main()` in the  `test.c` in `installPhase` is renamed to `__main_argc_argv()`. The latest force push returns this PR to 3.1.14 accordingly.

File: pkgs/development/compilers/emscripten/default.nix [hunk 1]
```diff
@@ -39,11 +39,6 @@ stdenv.mkDerivation rec {
       src = ./0001-emulate-clang-sysroot-include-logic.patch;
       resourceDir = "${llvmEnv}/lib/clang/${llvmPackages.release_version}/";
     })
-    (fetchpatch {
```
[the remark this answers is not in the document]
Comment (willcohen) [reply pkgs/development/compilers/emscripten/default.nix:42]: Sorry for missing. Adjusted with latest force push.

**Comment by willcohen** [comment]
On x86_darwin, `binaryen` fails with:

```
clang-11: warning: argument unused during compilation: '-L/private/tmp/nix-build-binaryen-109.drv-0/source/build/lib' [-Wunused-command-line-argument]
/private/tmp/nix-build-binaryen-109.drv-0/source/test/example/c-api-relooper-unreachable-if.cpp:4:10: fatal error: 'cassert' file not found
#include <cassert>
         ^~~~~~~~~
1 error generated.
Traceback (most recent call last):
  File "/private/tmp/nix-build-binaryen-109.drv-0/source/build/../check.py", line 416, in <module>
    sys.exit(main())
  File "/private/tmp/nix-build-binaryen-109.drv-0/source/build/../check.py", line 399, in main
    TEST_SUITES[test]()
  File "/private/tmp/nix-build-binaryen-109.drv-0/source/build/../check.py", line 301, in run_example_tests
    subprocess.check_call(extra)
  File "/nix/store/s3xwz0p7ng23dn2ixhksrdgs3gmq8fbg-python3-3.10.4/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['clang', '/private/tmp/nix-build-binaryen-109.drv-0/source/test/example/c-api-relooper-unreachable-if.cpp', '-c', '-o', 'example.o', '-I/private/tmp/nix-build-binaryen-109.drv-0/source/src', '-g', '-L/private/tmp/nix-build-binaryen-109.drv-0/source/build/lib', '-pthread', '-std=c++17']' returned non-zero exit status 1.
error: builder for '/nix/store/b8lb75mmf4vxdrk8718bxpvy10x84wna-binaryen-109.drv' failed with exit code 1;
       last 10 log lines:
       > Traceback (most recent call last):
       >   File "/private/tmp/nix-build-binaryen-109.drv-0/source/build/../check.py", line 416, in <module>
       >     sys.exit(main())
       >   File "/private/tmp/nix-build-binaryen-109.drv-0/source/build/../check.py", line 399, in main
       >     TEST_SUITES[test]()
       >   File "/private/tmp/nix-build-binaryen-109.drv-0/source/build/../check.py", line 301, in run_example_tests
       >     subprocess.check_call(extra)
       >   File "/nix/store/s3xwz0p7ng23dn2ixhksrdgs3gmq8fbg-python3-3.10.4/lib/python3.10/subprocess.py", line 369, in check_call
       >     raise CalledProcessError(retcode, cmd)
       > subprocess.CalledProcessError: Command '['clang', '/private/tmp/nix-build-binaryen-109.drv-0/source/test/example/c-api-relooper-unreachable-if.cpp', '-c', '-o', 'example.o', '-I/private/tmp/nix-build-binaryen-109.drv-0/source/src', '-g', '-L/private/tmp/nix-build-binaryen-109.drv-0/source/build/lib', '-pthread', '-std=c++17']' returned non-zero exit status 1.
       For full logs, run 'nix log /nix/store/b8lb75mmf4vxdrk8718bxpvy10x84wna-binaryen-109.drv'.
```

**Comment by willcohen** [comment]
As noted in #175358, `binaryen` fails when tests are enabled on darwin, which is fixed with the latest push. My error in rebasing was to only remove the `broken` on darwin tag, and not then re-disable tests for darwin too.

**Comment by willcohen** [comment]
Result of `nixpkgs-review pr 178620` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>open-music-kontrollers.mephisto</li>
    <li>tinygo</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>binaryen</li>
    <li>emscripten</li>
    <li>python310Packages.filecheck</li>
    <li>python39Packages.filecheck</li>
  </ul>
</details>

**Review by SuperSandro2000 — APPROVED** [review]
LGTM

**Comment by willcohen** [comment]
One more change I'd like to make before merge. In testing this PR's release on some test projects, I noticed that https://github.com/WebAssembly/binaryen/pull/4321 fixed an issue I kept running into, which is now merged upstream. I'd like to patch that in too!

**Comment by willcohen** [comment]
Thank you!

**Comment by willcohen** [comment]
Followup version bump in #181362.

**Pull request merged**